pkg/cfn/builder/api.go

```diff
@@ -0,0 +1,210 @@
+}
+
+func makeSlice(i ...*gfn.StringIntrinsic) []*gfn.StringIntrinsic {
+	return i
+}
+
+func makeStringSlice(s ...string) []*gfn.StringIntrinsic {
+	slice := []*gfn.StringIntrinsic{}
+	for _, i := range s {
+		slice = append(slice, gfn.NewString(i))
+	}
+	return slice
+}
+
+func makeAutoNameTag(suffix string) gfn.Tag {
+	return gfn.Tag{
+		Key:   gfn.NewString("Name"),
+		Value: makeSub(fmt.Sprintf("${%s}/%s", awsStackName, suffix)),
+	}
+}
+
+var refStackName = makeRef(awsStackName)
+
+func (r *resourceSet) newParameter(name, valueType, defaultValue string) *gfn.StringIntrinsic {
+	p := map[string]string{"Type": valueType}
+	if defaultValue != "" {
+		p["Default"] = defaultValue
+	}
+	r.template.Parameters[name] = p
+	return makeRef(name)
+}
+
+func (r *resourceSet) newStringParameter(name, defaultValue string) *gfn.StringIntrinsic {
+	return r.newParameter(name, "String", defaultValue)
+}
+
+func (r *resourceSet) newNumberParameter(name, defaultValue string) *gfn.StringIntrinsic {
+	return r.newParameter(name, "Number", defaultValue)
+}
+
+func (r *resourceSet) newResource(name string, resource interface{}) *gfn.StringIntrinsic {
+	r.template.Resources[name] = resource
+	e := reflect.ValueOf(resource).Elem()
+	if e.Kind() == reflect.Struct {
+		f := e.FieldByName("Tags")
+		if f.IsValid() && f.CanSet() {
+			tag := reflect.ValueOf(makeAutoNameTag(name))
+			if f.Type() == reflect.ValueOf([]gfn.Tag{}).Type() {
+				f.Set(reflect.Append(f, tag))
+			}
+		}
+	}
+	return makeRef(name)
+}
+
+func (r *resourceSet) renderJSON() ([]byte, error) {
+	return r.template.JSON()
+}
+
+func exportName(prefix, output string) string {
+	return fmt.Sprintf("${%s}::%s", prefix, output)
+}
+
+func (r *resourceSet) newOutput(name string, value interface{}, export bool) {
+	o := map[string]interface{}{"Value": value}
+	if export {
+		o["Export"] = map[string]map[string]string{
+			"Name": map[string]string{fnSub: exportName(awsStackName, name)},
+		}
+	}
+	r.template.Outputs[name] = o
+	r.outputs = append(r.outputs, name)
+}
+
+func (r *resourceSet) newJoinedOutput(name string, value []*gfn.StringIntrinsic, export bool) {
+	r.newOutput(name, map[string][]interface{}{fnJoin: []interface{}{",", value}}, export)
+}
+
+func (r *resourceSet) newOutputFromAtt(name, att string, export bool) {
+	r.newOutput(name, map[string]string{fnGetAtt: att}, export)
+}
+
+func makeImportValue(prefix, output string) *gfn.StringIntrinsic {
+	return gfn.NewStringIntrinsic(fnImportValue, makeSub(exportName(prefix, output)))
+}
+
+func getOutput(stack *cfn.Stack, key string) *string {
+	for _, x := range stack.Outputs {
+		if *x.OutputKey == key {
+			return x.OutputValue
+		}
+	}
+	return nil
+}
+
+func setOutput(obj interface{}, key, value string) error {
+	e := reflect.ValueOf(obj).Elem()
+	if e.Kind() == reflect.Struct {
+		f := e.FieldByName(key)
+		if f.IsValid() && f.CanSet() {
+			switch f.Kind() {
+			case reflect.String:
+				f.SetString(value)
+			case reflect.Slice:
```

Comment:
It feels like there are few too many switches. Perhaps I should think about how to make it more readable also.